homeassistant/components/automation/__init__.py

```diff
@@ -700,73 +709,132 @@ async def _async_process_config(
-                automation_id,
-                name,
-                config_block[CONF_TRIGGER],
-                cond_func,
-                action_script,
-                initial_state,
-                variables,
-                config_block.get(CONF_TRIGGER_VARIABLES),
-                raw_config,
-                raw_blueprint_inputs,
-                config_block[CONF_TRACE],
+        automation_id: str | None = config_block.get(CONF_ID)
+        name: str = config_block.get(CONF_ALIAS) or f"{config_key} {list_no}"
+
+        initial_state: bool | None = config_block.get(CONF_INITIAL_STATE)
+
+        action_script = Script(
+            hass,
+            config_block[CONF_ACTION],
+            name,
+            DOMAIN,
+            running_description="automation actions",
+            script_mode=config_block[CONF_MODE],
+            max_runs=config_block[CONF_MAX],
+            max_exceeded=config_block[CONF_MAX_EXCEEDED],
+            logger=LOGGER,
+            # We don't pass variables here
+            # Automation will already render them to use them in the condition
+            # and so will pass them on to the script.
+        )
+
+        if CONF_CONDITION in config_block:
+            cond_func = await _async_process_if(hass, name, config_block)
+
+            if cond_func is None:
+                continue
+        else:
+            cond_func = None
+
+        # Add trigger variables to variables
+        variables = None
+        if CONF_TRIGGER_VARIABLES in config_block:
+            variables = ScriptVariables(
+                dict(config_block[CONF_TRIGGER_VARIABLES].as_dict())
             )
+        if CONF_VARIABLES in config_block:
+            if variables:
+                variables.variables.update(config_block[CONF_VARIABLES].as_dict())
+            else:
+                variables = config_block[CONF_VARIABLES]
+
+        entity = AutomationEntity(
+            automation_id,
+            name,
+            config_block[CONF_TRIGGER],
+            cond_func,
+            action_script,
+            initial_state,
+            variables,
+            config_block.get(CONF_TRIGGER_VARIABLES),
+            automation_config.raw_config,
+            automation_config.raw_blueprint_inputs,
+            config_block[CONF_TRACE],
+        )
+        entities.append(entity)
+
+    return entities
+
+
+async def _async_process_config(
+    hass: HomeAssistant,
+    config: dict[str, Any],
+    component: EntityComponent[AutomationEntity],
+) -> bool:
+    """Process config and add automations.
+
+    Returns if blueprints were used.
+    """
+
+    def find_match(
+        automations: list[AutomationEntity],
+        automation_configs: list[AutomationEntityConfig],
+    ) -> tuple[int, int] | None:
+        """Return indices of first automation and matching configuration."""
+        for automation_idx, automation in enumerate(automations):
+            for config_idx, config in enumerate(automation_configs):
```

Comment:
Can we do better than this? This is wildly inefficient.

Reply:
In conjunction with the outer loop, this is actually O(N^3) in the worst case.

Consider if a person has 20 automations and they just changed the first 10 all at once. (Perhaps by editing a value pulled from secrets.yaml). The last 10 remain the same. That would result in this potentially expensive equality check running 1650 times (10*20+10*19+10*18+10*17+10*16+10*15+10*14+10*13+10*12+10*11+10*10,  each value being the number of times the inner comparison is run for each call of this method).

If it was 40 automations and they changed the first 20 at once, it would hit the comparison 12800 times!

Reply:
Rewritten to only do one pass, to avoid testing the same non-matches repeatedly